server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,221 @@
+        return masters.stream().filter(Objects::nonNull).collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns the set of distinct non-null master nodes seen in this history.
+     * @return The set of all non-null master nodes seen. Could be empty
+     */
+    public Set<DiscoveryNode> getDistinctMastersSeen() {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes();
+        return masterHistoryCopy.stream().map(timeAndMaster -> timeAndMaster.master).filter(Objects::nonNull).collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns true if a non-null master was seen at any point in the last n seconds, or if the last-seen master was more than n seconds
+     * ago and non-null.
+     * @param n The number of seconds to look back
+     * @return true if the current master is non-null or if a non-null master was seen in the last n seconds
+     */
+    public boolean hasSeenMasterInLastNSeconds(int n) {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes();
+        long now = nowSupplier.get();
+        TimeValue nSeconds = new TimeValue(n, TimeUnit.SECONDS);
+        long nSecondsAgo = now - nSeconds.getMillis();
+        return getCurrentMaster() != null
+            || masterHistoryCopy.stream().anyMatch(timeAndMaster -> timeAndMaster.time > nSecondsAgo && timeAndMaster.master != null);
+    }
+
+    /*
+     * This method creates a copy of masterHistory that only has entries from more than 30 minutes before now (but leaves the newest
+     * entry in even if it is more than 30 minutes old).
+     */
+    private List<TimeAndMaster> getMasterHistoryForLast30Minutes() {
+        List<TimeAndMaster> masterHistoryCopy = masterHistory; // In case masterHistory is swapped while we're using it
+        if (masterHistoryCopy.size() < 2) {
+            return masterHistoryCopy;
+        }
+        long now = nowSupplier.get();
+        TimeValue thirtyMinutes = new TimeValue(30, TimeUnit.MINUTES);
+        long thirtyMinutesAgo = now - thirtyMinutes.getMillis();
+        TimeAndMaster mostRecent = masterHistoryCopy.isEmpty() ? null : masterHistoryCopy.get(masterHistoryCopy.size() - 1);
+        masterHistoryCopy = masterHistoryCopy.stream()
+            .filter(timeAndMaster -> timeAndMaster.time > thirtyMinutesAgo)
+            .collect(Collectors.toList());
+        if (masterHistoryCopy.isEmpty() && mostRecent != null) { // The most recent entry was more than 30 minutes ago
+            masterHistoryCopy.add(mostRecent);
+        }
+        return masterHistoryCopy;
+    }
+
+    /**
+     * Clears out anything from masterHistory that is from more than 30 minutes before now (but leaves the newest entry in even if it is
+     * more than 30 minutes old). Rather than being scheduled, this method is called whenever the cluster state changes.
+     */
+    private void removeOldMasterHistory(List<TimeAndMaster> newMasterHistory) {
+        if (newMasterHistory.size() < 2) {
```

Comment:
> Are we looking for size ==1 here? Or is empty still something we want to deal with?

If you happen to hit this method before the very first ClusterChangedEvent then this will be empty. That would be incredibly hard to do, and it might even happen before the cluster will start accepting rest requests. But I figured better to protect against it than not since there's no additional cost.

> Could the list ever just have a `null` entry?

No.